File: src/main/java/runtime/judge/Judge.java
package runtime.judge;

import com.google.common.base.Preconditions;
import runtime.parser.Example;
import runtime.parser.ObjectReader;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.joining;

public class Judge {

	public record Result(long cost, boolean wrong_answer, String expect_output, String actual_output) {
		public boolean time_limit_exceeded() { return cost > 5_000; }
		public boolean ok() { return !wrong_answer && !time_limit_exceeded(); }
	}

	public static List<Result> run(Method method, List<Example> examples) {
		Preconditions.checkState(!Modifier.isStatic(method.getModifiers()));
		Preconditions.checkState(Modifier.isPublic(method.getModifiers()));

		List<Result> out = new ArrayList<>();
		for (Example example : examples) {
			long start = System.currentTimeMillis();
			Object returned = call(method, example.input());
			long cost = System.currentTimeMillis() - start;
			out.add(compare_result(method, example.output(), returned, cost));
		}
		return out;
	}

	private static Object call(Method method, String input) {
		// parse args
		ObjectReader args_reader = new ObjectReader(input);
		Object[] args = Arrays.stream(method.getParameterTypes())
				.map(args_reader::read_param)
				.toArray();
		Preconditions.checkState(args_reader.eof());

		// call
		try {
			Object instance = method.getDeclaringClass().getDeclaredConstructor().newInstance();
			return method.invoke(instance, args);
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	private static Result compare_result(Method method, String output, Object returned, long cost) {
		Object output_obj = new ObjectReader(output).read_return(method.getReturnType());
		boolean output_correct = Objects.equals(output_obj, returned);
		if (output_correct) {
			return new Result(cost, false, output, output);
		} else {
			return new Result(cost, true, output, format_obj(output_obj));
		}
	}
	private static String format_obj(Object obj) {
		if (obj.getClass().isArray()) {
			return IntStream.range(0, Array.getLength(obj))
					.mapToObj(i -> Array.get(obj, i))
					.map(Judge::format_obj)
					.collect(joining(",", "[", "]"));
		} else {
			return obj.toString();
		}
	}

}

File: src/main/scala/runtime/judge/Judge.scala
package runtime.judge

import runtime.ListNode
import runtime.parser.{Example, ObjectReader2}
import runtime.repo.MetaDataNode

import java.lang.reflect.{Method, Modifier}
import scala.collection.mutable.ArrayBuffer

case class JudgeResult(cost: Long, wrong_answer: Boolean, expect_output: String, actual_output: String) {
	def time_limit_exceeded: Boolean = cost > 5_000
	def ok: Boolean = !wrong_answer && !time_limit_exceeded
}

object Judge {

	def run(instance: Any, method: Method, example: Example, meta: MetaDataNode): JudgeResult = {
		assert(!Modifier.isStatic(method.getModifiers))
		assert(Modifier.isPublic(method.getModifiers))

		val (cost, actual_output) = call(instance, method, example.input, meta)
		example.output match {
			case Some(expect_output) => compare_result(meta, expect_output, actual_output, cost)
			case None => JudgeResult(cost, wrong_answer = false, "", "")
		}
	}

	private def call(instance: Any, method: Method, input: String, meta: MetaDataNode): (Long, Any) = {
		val reader = new ObjectReader2(input)
		val args: Array[Any] = meta.params.map(p => reader.read_param(p.`type`))
		assert(reader.eof)

		val start = System.currentTimeMillis()
		val returned = method.invoke(instance, args: _*)
		val cost = System.currentTimeMillis() - start
		(cost, returned)
	}

	private def compare_result(meta: MetaDataNode, expect_str: String, actual: Any, cost: Long): JudgeResult = {
		val simple_return_type = meta.`return` == null
		val expect = new ObjectReader2(expect_str).read_return(meta.`return`.`type`, simple_return_type)
		val wrong_answer = !is_equal(expect, actual)
		JudgeResult(cost, wrong_answer, expect_str, format_obj(actual))
	}
	private def is_equal(a: Any, b: Any): Boolean = {
		if (a == b) return true

		if (a != null && a.getClass.isArray) {
			b.getClass.isArray && a.asInstanceOf[Array[_]].zip(b.asInstanceOf[Array[_]]).forall { case (a, b) =>
				is_equal(a, b)
			}
		} else if (a.isInstanceOf[ListNode]) {
			b.isInstanceOf[ListNode] && is_list_node_equal(a.asInstanceOf[ListNode], b.asInstanceOf[ListNode])
		} else {
			false
		}
	}
	private def is_list_node_equal(a: ListNode, b: ListNode): Boolean = {
		if (a == null && b == null) return true
		if (a == null || b == null) return false
		if (a.x != b.x) return false
		is_list_node_equal(a.next, b.next)
	}
	private def format_obj(obj: Any): String = {
		obj match {
			case arr: Array[_] => arr.map(format_obj).mkString("[", ",", "]")
			case list: List[_] => list.map(format_obj).mkString("[", ",", "]")
			case list: ListNode => {
				val out = ArrayBuffer.empty[Int]
				var cur = list
				while (cur != null) {
					out += cur.x
					cur = cur.next
				}
				out.mkString("[", ",", "]")
			}
			case _ => String.valueOf(obj)
		}
	}

}
